### PR TITLE
Add support for QuantumCircuit.layout to qpy

### LINKED ISSUE #10112: QPy roundtripping discards physical qubit information on circuit
### Environment

- **Qiskit Terra version**: main
- **Python version**: 3.9
- **Operating system**: OSX

### What is happening?

When roundtripping a transpiled circuit for a backend, and then dumping to OpenQASM 3 the circuit is no longer on physical qubits when emitted to OpenQASM 3. This impacts the backend when circuit merging is not applied.

### How can we reproduce the issue?

```python
from qiskit import qasm3, transpile, QuantumCircuit, ClassicalRegister, QuantumRegister
from qiskit import qpy

qr = QuantumRegister(1)
cr = ClassicalRegister(1)
qc = QuantumCircuit(1, 1)
qc.x(0)

qc_transpiled = transpile(qc, backend)

with open('transpiled.qpy', 'wb') as fd:
    qpy.dump(qc_transpiled, fd)

with open('transpiled.qpy', 'rb') as fd:
    qc_transpiled = qpy.load(fd)[0]

print(qc_qasm3 := qasm3.dumps(qc_transpiled, disable_constants=True, experimental=qasm3.ExperimentalFeatures.SWITCH_CASE_V1))
```

Emits
```
OPENQASM 3;
include "stdgates.inc";
bit[1] c;
qubit[7] _all_qubits;
let q = _all_qubits[0:6];
x q[0];
```

### What should happen?

Commenting out the QPy roundtripping

```
from qiskit import qasm3, transpile, QuantumCircuit, ClassicalRegister, QuantumRegister
from qiskit import qpy

qr = QuantumRegister(1)
cr = ClassicalRegister(1)
qc = QuantumCircuit(1, 1)
qc.x(0)

qc_transpiled = transpile(qc, backend)

#with open('transpiled.qpy', 'wb') as fd:
#    qpy.dump(qc_transpiled, fd)

#with open('transpiled.qpy', 'rb') as fd:
#    qc_transpiled = qpy.load(fd)[0]

print(qc_qasm3 := qasm3.dumps(qc_transpiled, disable_constants=True, experimental=qasm3.ExperimentalFeatures.SWITCH_CASE_V1))
```

Gives 
```
OPENQASM 3;
include "stdgates.inc";
bit[1] c;
x $0;
```

### Any suggestions?

This may be fixed by setting the layout on roundtripping

```python
from qiskit import qasm3, transpile, QuantumCircuit, ClassicalRegister, QuantumRegister
from qiskit import qpy
from qiskit.transpiler import Layout

qr = QuantumRegister(1)
cr = ClassicalRegister(1)
qc = QuantumCircuit(1, 1)
qc.x(0)

qc_transpiled = transpile(qc, backend)

with open('transpiled.qpy', 'wb') as fd:
    qpy.dump(qc_transpiled, fd)

with open('transpiled.qpy', 'rb') as fd:
    qc_transpiled = qpy.load(fd)[0]


qc_transpiled._layout = Layout({qubit: idx for idx, qubit in enumerate(qc_transpiled.qubits)})

print(qc_qasm3 := qasm3.dumps(qc_transpiled, disable_constants=True, experimental=qasm3.ExperimentalFeatures.SWITCH_CASE_V1))
```

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds the missing support for QuantumCircuit.layout to the qpy format. This necessitates bumping the QPY format version to 8 to accommodate the extra data needed for representing the details of the layout. The tricky piece with representing the 3 TranspileLayout attributes is representing the virtual bits in the initial layout because there is no guarantee that the input circuit's registers are in the output circuit (typically they are not when transpile() is used).

### Details and comments

TODO:

- [x] Finish documentation
- [x] Add backwards compatibility tests
- [x] ~Add test coverage for layout with standalone qubits (no associated register)~
- [x] Add release note

Fixes #10112

### CODE REVIEW

File: qiskit/qpy/__init__.py [hunk 1]
```diff
@@ -126,6 +126,62 @@
 by ``num_circuits`` in the file header). There is no padding between the
 circuits in the data.
 
+.. _qpy_version_8:
+
+Version 8
+=========
+
+Version 8 adds support for handling a :class:`~.TranspileLayout` stored in the
+:attr:`.QuantumCircuit.layout` attribute. In version 8 imeediately following the
```
Comment (mtreinish) [inline qiskit/qpy/__init__.py:135]:
```suggestion
:attr:`.QuantumCircuit.layout` attribute. In version 8 immediately following the
```

File: qiskit/qpy/__init__.py [hunk 2]
```diff
@@ -126,6 +126,62 @@
...
+        int32_t initial_layout_size;
+        int32_t input_mapping_size;
+        int32_t final_layout_size;
+        uint32_t extra_registers;
+    }
+
+If any of the signed values are ``-1`` this indicates the corresponding
+attribute is ``None``
+
+Imeediately following the ``LAYOUT`` struct there is a :ref:`qpy_registers` struct
```
Comment (mtreinish) [inline qiskit/qpy/__init__.py:155]:
```suggestion
Immediately following the ``LAYOUT`` struct there is a :ref:`qpy_registers` struct
```

File: qiskit/qpy/__init__.py [hunk 3]
```diff
@@ -126,6 +126,62 @@
...
+    }
+
+Where a value of ``-1`` indicates ``None`` (as in no register is associated
+with the bit). Following each ``INITIAL_LAYOUT_BIT`` struct is ``register_size``
+bytes for a ``utf8`` encoded string for the register name.
+
+Following the initial layout there is ``input_mapping_size`` array of
+``uint32_t`` integers representing the positions of the phyiscal bit from the
+initial layout. This enbables constructing a list of virtual bits where the
+array index is it's input mapping position.
```
Comment (jlapeyre) [inline qiskit/qpy/__init__.py:178]:
```suggestion
array index is its input mapping position.
```
eternal foible of English

**Review by jakelishman — COMMENTED** [review]
The format of the new data looks pretty clean to me, which is always nice to see when we're dealing with our register nonsense.  I think there's a bit of scope for improving the test coverage (esp in the backwards compatibility tests), but implementation-wise, I'm happy.

File: qiskit/qpy/__init__.py [hunk 4]
```diff
@@ -126,6 +126,62 @@
...
+------
+
+.. code-block:: c
+
+    struct {
+        int32_t initial_layout_size;
+        int32_t input_mapping_size;
+        int32_t final_layout_size;
+        uint32_t extra_registers;
+    }
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/qpy/__init__.py:150]: Ah, yeah I added that after I started writing the docs because as I was testing locally I needed to differentiate between no layout and an empty layout. I'll update this to reflect that.

As for the use of `_Bool` I guess yeah that's true it is 1 byte long, I took the type from: https://docs.python.org/3/library/struct.html#format-characters but I neglected to read the footnotes about `_Bool` a more accurate type is probably `char` based on that. I think we can go through and update the docs to better reflect that in a follow up PR.

File: test/python/qpy/test_circuit_load_from_qpy.py [hunk 5]
```diff
@@ -35,6 +36,8 @@ def assert_roundtrip_equal(self, circuit):
         new_circuit = load(qpy_file)[0]
 
         self.assertEqual(circuit, new_circuit)
+        if getattr(circuit, "layout", None):
+            self.assertEqual(circuit.layout, new_circuit.layout)
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply test/python/qpy/test_circuit_load_from_qpy.py:40]: I think there was a reason I did it I don't remember exactly why though. I'll try dropping it and see what happens

File: test/python/qpy/test_circuit_load_from_qpy.py [hunk 6]
```diff
@@ -67,3 +70,67 @@ def test_rzx_calibration_echo(self, angle):
...
+    def test_transpile_layout(self, opt_level):
+        """Test layout preserved after transpile."""
+        qc = QuantumCircuit(2)
+        qc.h(0)
+        qc.cx(0, 1)
+        qc.measure_all()
+        backend = FakeSherbrooke()
+        tqc = transpile(qc, backend, optimization_level=opt_level)
+        self.assert_roundtrip_equal(tqc)
+
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply test/python/qpy/test_circuit_load_from_qpy.py:89]: That's fair, I'll add 2 tests one which will require routing to be run and also one with an explicit `None` for final layout. I also should add a test with an empty transpiler layout too.

File: qiskit/qpy/binary_io/circuits.py [hunk 7]
```diff
@@ -761,6 +763,143 @@ def _write_registers(file_obj, in_circ_regs, full_bits):
...
+            reg_size = len(reg_name_bytes)
+        file_obj.write(
+            struct.pack(
+                formats.INITIAL_LAYOUT_BIT_PACK,
+                index,
+                reg_size,
+            )
+        )
+        if reg_name_bytes is not None:
+            file_obj.write(reg_name_bytes)
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/qpy/binary_io/circuits.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10148/commits/07fcec9200f688abf02a8b36ef96f26771f5bcb5

File: qiskit/qpy/__init__.py [hunk 8]
```diff
@@ -126,6 +126,62 @@
...
+
+If any of the signed values are ``-1`` this indicates the corresponding
+attribute is ``None``
+
+Immediately following the ``LAYOUT`` struct there is a :ref:`qpy_registers` struct
+for ``extra_registers`` (specifically the format introduced in :ref:`qpy_version_4`
+standalone register definitions that aren't present in the circuit. Then there
+are ``initial_layout_size`` ``INITIAL_LAYOUT_BIT`` structs to define the
+:attr:`.TranspileLayout.initial_layout` attribute.
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/qpy/__init__.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10148/commits/dd0efc149d73525ec53ca60681a89e617fb5fe84

File: test/python/qpy/test_circuit_load_from_qpy.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply test/python/qpy/test_circuit_load_from_qpy.py:89]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10148/commits/f67ded457f57251b4f5f44d0b58d5bae44b24bff

File: qiskit/qpy/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/qpy/__init__.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10148/commits/dd0efc149d73525ec53ca60681a89e617fb5fe84

File: test/python/qpy/test_circuit_load_from_qpy.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply test/python/qpy/test_circuit_load_from_qpy.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10148/commits/f67ded457f57251b4f5f44d0b58d5bae44b24bff

File: test/qpy_compat/test_qpy.py [hunk 9]
```diff
@@ -574,6 +574,24 @@ def generate_open_controlled_gates():
...
+    qc.swap(0, 1)
+    qc.cx(0, 2)
+    qc._layout = TranspileLayout(
+        Layout({bit: index for index, bit in enumerate(qc.qubits)}),
+        input_qubit_mapping={bit: index for index, bit in enumerate(qc.qubits)},
+        final_layout=Layout.from_qubit_list([qc.qubits[1], qc.qubits[0], qc.qubits[2]]),
+    )
+    return [qc]
+
+
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply test/qpy_compat/test_qpy.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10148/commits/af4e8272f5a1d306ff931711fd140362edff49fe

File: test/qpy_compat/test_qpy.py [hunk 10]
```diff
@@ -651,6 +671,15 @@ def assert_equal(reference, qpy, count, version_parts, bind=None):
...
                 sys.exit(1)
 
+    if (
+        version_parts > (0, 24, 0)
+        and isinstance(reference, QuantumCircuit)
+        and reference.layout != qpy.layout
+    ):
+        msg = f"Circuit {count} layout mismatch {reference.layout} != {qpy.layout}"
+        sys.stderr.write(msg)
+        sys.exit(4)
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply test/qpy_compat/test_qpy.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10148/commits/659a391ef4d4051495e6fff8a6c194488bcb7a09

File: test/python/qpy/test_circuit_load_from_qpy.py [hunk 11]
```diff
@@ -35,6 +36,8 @@ def assert_roundtrip_equal(self, circuit):
         new_circuit = load(qpy_file)[0]
 
         self.assertEqual(circuit, new_circuit)
+        if circuit.layout is not None:
+            self.assertEqual(circuit.layout, new_circuit.layout)
```
Comment (jakelishman) [inline test/python/qpy/test_circuit_load_from_qpy.py:40]: I'm pretty sure this assertion should be valid even if `circuit.layout` _is_ None.

**Review by jakelishman — APPROVED** [review]
One mostly nit-pick comment, but otherwise I'm still happy with the state of this.

#10392 will need updating after this PR merges to fix documentation conflicts only, but that's not a big deal.

File: test/python/qpy/test_circuit_load_from_qpy.py (hunk 11 above)
Comment (mtreinish) [reply test/python/qpy/test_circuit_load_from_qpy.py:40]: That's fair, I think I did this because I didn't want to assert the layout in cases when we were calling `transpile()` and not adding a layout or something like that, but it's been long enough I don't really remember. But, yeah if circuit.layout is None then this should be true too.

```suggestion
        self.assertEqual(circuit, new_circuit)
        self.assertEqual(circuit.layout, new_circuit.layout)
```

**Review by jakelishman — APPROVED** [review]
